### PR TITLE
Initial implementation CI environments

### PR DESCRIPTION
**Pull request by ss-gxp**
The environment can include several branches.
When commit to a branch, builds all environments in which the branch enters.
When build the environment, the branches related to it merge into it.
If no environment is specified, it should work as before.

Builds:
ci-env-1

Build view:
ci-env-2

Edit project
ci-env-3

### CODE REVIEW

**Comment by corpsee** [comment]
Excellent work! Thank you :beer:. I will look at this feature soon.
Can you add the short documentation (for environments configuration format) and make rebase on current master branch?

**Comment by corpsee** [comment]
And it would be great if you could fix the tests according your changes. See https://travis-ci.org/corpsee/php-censor/jobs/214704570

**Comment by bochkovprivate** [comment]
Useful feature.

I have a number of solutions in my project. Solutions differ with each other by the database structure and the set of modules included.
Each time I submit a new commit I run tests against all solutions.

I could probably use this environments if I had an access to environment name variable in .php-censor.yml.
For example:
```
setup:
    mysql:
        - "CREATE DATABASE rfm_db_test_%ENVIRONMENT%_%BUILD%;"
# fill db with data
# ...
# run some tests
complete:
    mysql:
        - "DROP DATABASE IF EXISTS rfm_db_test_%ENVIRONMENT%_%BUILD%;"
```

**Comment by corpsee** [comment]
I think it is good idea - seting the %ENVIRONMENT% variable if the environment is set. @ss-gxp What are you think about it?

**Comment by ss-gxp** [comment]
Yes, `%ENVIRONMENT%` variable already exists.
I will fix the tests, write documentation and do rebase (do with force push or just merge the upstream? I usually do not use force push, but history will be cleaner).

**Comment by corpsee** [comment]
Docs :heart: Thanks!

File: src/PHPCensor/Migrations/20170321131931_add_environment.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        if (!$this->hasTable('environment')) {
+            $table->create();
+        }
+
+        if (!$table->hasColumn('project_id')) {
+            $table->addColumn('project_id', 'integer')->save();
+        }
+
+        if (!$table->hasColumn('name')) {
+            $table->addColumn('name', 'string', ['limit' => 20])->save();
```
Comment (corpsee) [inline src/PHPCensor/Migrations/20170321131931_add_environment.php]: I would make a limit of 250 (like a 'branch' field in the 'build' table). 20 is too short

File: src/PHPCensor/Migrations/20170321131931_add_environment.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        if (!$table->hasIndex(['project_id', 'name'])) {
+            $table->addIndex(['project_id', 'name'])->save();
+        }
+
+        $table = $this->table('build');
+
+        if (!$table->hasColumn('environment')) {
+            $table->addColumn('environment', 'string', ['limit' => 20])->save();
```
Comment (corpsee) [inline src/PHPCensor/Migrations/20170321131931_add_environment.php]: I would make a limit of 250 (like a 'branch' field in the 'build' table). 20 is too short

File: src/PHPCensor/Helper/BuildInterpolator.php [hunk 3]
```diff
@@ -37,6 +37,7 @@ public function setupInterpolationVars(BaseBuild $build, $buildPath, $url)
         $this->interpolation_vars['%COMMIT_URI%'] = $build->getCommitLink();
         $this->interpolation_vars['%BRANCH%'] = $build->getBranch();
         $this->interpolation_vars['%BRANCH_URI%'] = $build->getBranchLink();
+        $this->interpolation_vars['%ENVIRONMENT%'] = $build->getEnvironment();
```
Comment (corpsee) [inline src/PHPCensor/Helper/BuildInterpolator.php]: I would add:
```php
putenv('PHPCI_ENVIRONMENT=' . $this->interpolation_vars['%ENVIRONMENT%']);
```

File: src/PHPCensor/Model/Build.php [hunk 4]
```diff
@@ -165,6 +168,11 @@ class Build extends Model
             'nullable' => true,
             'default'  => null,
         ],
+        'environment' => [
+            'type'    => 'varchar',
+            'length'  => 20,
```
Comment (corpsee) [inline src/PHPCensor/Model/Build.php]: I would make a limit of 250 (like a 'branch' field in the 'build' table). 20 is too short

**Comment by corpsee** [comment]
Thanks! I merged manualy this branch.

**Pull request closed without merging**